lib/srv/db/mysql/autousers.go

```diff
@@ -0,0 +1,367 @@
+/*
+Copyright 2023 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package mysql
+
+import (
+	"context"
+	_ "embed"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"hash/fnv"
+	"strings"
+
+	"github.com/go-mysql-org/go-mysql/client"
+	"github.com/gravitational/trace"
+	"golang.org/x/exp/slices"
+
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/srv/db/common"
+)
+
+// activateUserDetails contains details about the user activation request and
+// will be marshaled into a JSON parameter for the stored procedure.
+type activateUserDetails struct {
+	// Roles is a list of roles to be assigned to the user.
+	//
+	// MySQL stored procedure does not accept array of strings thus using a
+	// JSON to bypass this limit.
+	Roles []string `json:"roles"`
+	// AuthOptions specifies auth options like "IDENTIFIED xxx" used when
+	// creating a new user.
+	//
+	// Using a JSON string can bypass VARCHAR character limit.
+	AuthOptions string `json:"auth_options"`
+	// Attributes specifies user attributes used when creating a new user.
+	//
+	// User attributes is a MySQL JSON in MySQL databases.
+	//
+	// To check current user's attribute:
+	// SELECT * FROM INFORMATION_SCHEMA.USER_ATTRIBUTES WHERE CONCAT(USER, '@', HOST) = current_user()
+	//
+	// Reference:
+	// https://dev.mysql.com/doc/refman/8.0/en/information-schema-user-attributes-table.html
+	Attributes struct {
+		// User is the original Teleport user name.
+		//
+		// Find a Teleport user (with "admin" privilege):
+		// SELECT * FROM INFORMATION_SCHEMA.USER_ATTRIBUTES WHERE ATTRIBUTE->"$.user" = "teleport-user-name";
+		User string `json:"user"`
+	} `json:"attributes"`
+}
+
+// ActivateUser creates or enables the database user.
+func (e *Engine) ActivateUser(ctx context.Context, sessionCtx *common.Session) error {
+	if sessionCtx.Database.GetAdminUser() == "" {
```

Comment:
I think this bug only happens when `AdminUser` is provided, but `name` is empty:
```
  databases:
  - name: "self-hosted-mysql"
    protocol: "mysql"
    uri: "localhost:3306"
    static_labels:
      "teleport.dev/db-admin": "teleport-admin"
    admin_user:
      # nothing
```
Should we fallback to the label?

Reply:
> For auto-discovered, AdminUser is always nil thus using the fallback.

Let me check this agian because I think that in my local setup I have encounter a behavior where AdminUser!= nil and Name was empty. 

Should we fallback to the label?

> I think this bug only happens when AdminUser is provided, but Name is empty

If that is true then the logic seems to be fine.